reader-privacy.md

```diff
@@ -0,0 +1,242 @@
+passive observers from tracking user's actions as described above. It will also be a first step towards
+fully private IPNI protocol that will eliminate indexers as centralised observers.  
+
+### Non Goals
+
+* Writer, i.e. content provider or publisher, Privacy, which will be done in a separate specification
+* Retrieval Privacy, which is out of scope for the content routing subsystem.
+​
+## Background
+​
+Network indexers build their indexes by ingesting chains of Advertisements. Advertisement is a
+construct that allows Storage Providers to publish their CIDs in bulk (FIL deals) instead of doing
+that individually for each CID. A group of CIDs is represented by a unique ContextID as can be seen
+on the diagram below:
+​
+![Index building flow](resources/readers-privacy-1.png)
+
+## Specification
+​
+This specification improves the reader privacy by proposing changes to the Step 3, depicted above, where the client 
+supplies the content CID directly in order to lookup its corresponding providers.
+
+In order to protect the reader's privacy the proposal changes the way CID lookup works to the following:
+
+* A client who wants to do a lookup will calculate a hash over the CID's multihash (`hash(MH)`) and use it for the
+lookup query (hence the name double hashing);
+* In response to the hashed find request, the indexer will return a set of encrypted `ProviderRecordKey`s. 
+`ProviderRecordKey` will consist of the `peerID` concatenated with `contextID`. 
+Each `ProviderRecordKey` will be encrypted with a key derived from the *original* multihash value: 
+`enc(peerID || contextID, MH)`, where `hash` is a hash over the value, and `||` is concatenation 
+and `enc` is encryption over the value. In order to make sense of that payload, a passive observer would need 
+to get hold of the original CID that isn't revealed during the communication round;
+* Using the original multihash, the client will decrypt `ProviderRecordKey`s and then calculate a hash
+over the decrypted `peerID` part of it. Using such hash for each `ProviderRecordKey` the client would do another lookup 
+to get an encrypted `ProviderRecord` in response. `ProviderRecord` will contain information about the provider, 
+consisting of its *peerID* and *multiaddrs*. Each `ProviderRecord` will be encrypted 
+with a key derived from `peerID`. In order to make sense of that payload, a passive observer would need to 
+get hold of the original `peerID` that isn't revealed during the communication round;
+* Using a key derived from `peerID`, the client will decrypt `ProviderRecord`s and then reach out to the 
+provider directly to fetch the desired content. 
+* The client might choose to fetch additional `Metadata` that is supplied to IPNI in Advertisements.
+That will require another lookup by `hash(ProviderRecordKey)` to get `enc(Metadata, ProviderRecordKey)` in response. 
+*This step will not be required for IPFS as Bitswap protocol is assumed implicitly.*
```

Comment:
I'm confused by what you mean here (aside from the general IPFS == Bitswap thing which is likely to confuse others).

Do you mean that implementations that rely on libp2p protocol negotiation with multistream can choose to ignore the metadata request because they can just talk to the target peer directly (e.g. true for Bitswap, vanilla GraphSync, etc.)?

Reply:
Good point - removed that sentence. Let me explain myself.  

The client must be able to understand what protocol to use to speak to the returned providers. As all data on the IPNI server is encrypted - including metadata that defines the protocol - the server can not apply any filtering logic, as that would require knowledge of the original multihash which the client doesn't want to reveal. 

Having said that - filtering logic would have to be applied on the client side either by querying metadata or by doing libp2p negotiation.